DM-31722: Clean up utility code usage

timj opened the pull request:
Requires lsst/utils#100

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

timj committed:
Use lsst.utils utility code

Many utility routines have been relocated to the lsst.utils package.
Remove them from daf_butler and use them.

timj committed:
Use VERBOSE logging from lsst.utils

timj committed:
Remove utils.findFileResources

It is not used and ButlerURI version is preferred.

timj committed:
Remove safeMakeDir

os.makedirs is now safe

timj committed:
Fix mypy warning in logging

timj committed:
Switch to doImportType and add additional class checks

mypy noted that doImport can return a Module and this was not
expected in many cases. mypy also requires that there is a check
on the return value from doImportType to ensure that it is
of the correct class.

timj committed:
Rearrange isplit logic for mypy

Mypy does not like

  isplit(str, str | bytes)

so rearrange the logic to have distinct calls to isplit for
str and bytes.

timj committed:
Teach mypy linter about lsst.utils

timj committed:
Fix VERBOSE importing in cliLog

timj committed:
Move Singleton, immutable, and cached_getter to lsst.utils

TallJimbo commented on python/lsst/daf/butler/core/dimensions/_skypix.py:
```diff
@@ -229,7 +229,11 @@ def hasDependenciesIn(self, others: AbstractSet[str]) -> bool:
         # Docstring inherited from DimensionConstructionVisitor.
-        PixelizationClass = doImport(self._pixelizationClassName)
+        pixelization_class = doImportType(self._pixelizationClassName)
+        if not hasattr(pixelization_class, "MAX_LEVEL"):
```
I don't think we want this check (at least not here), as it makes it impossible to use a `PixelizationClass` that lacks `MAX_LEVEL` even if you've set the `max_level` dimension config explicitly.

timj replied:
Okay. I will have to rearrange the logic here so that it doesn't use that one-line if statement and only does the mypy check if it's actually going to call MAX_LEVEL.

timj replied:
I've reorganized it. Oddly mypy is not complaining that I have not asserted that this is a `lsst.sphgeom.Pixelization` when it's put into the `SkyPixSystem`.

TallJimbo commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -416,10 +412,17 @@ def from_raw(cls, serialized: Union[str, bytes]) -> ButlerLogRecords:
-        records = [ButlerLogRecord.parse_raw(line) for line in isplit(serialized, newline)  # type: ignore
-                   if line]
+        # argument to isplit() [which can't have two different types
+        # simultaneously] so we have to duplicate some logic.
```
I'm curious what the mypy error was with the original code.

timj replied:
It was complaining that Union[str, bytes] for the separator can't work for a `str` parameter that is being split because the `T` type can't then work out what the return type is. With the rewrite it knows that serialized is the same type as newline so can know what isplit returns. Fundamentally if a variable at any point has a Union type it can't be converted to a non-Union type without being assigned to a new variable.

TallJimbo replied:
> Fundamentally if a variable at any point has a Union type it can't be converted to a non-Union type without being assigned to a new variable.

I've definitely seen mypy "collapse" a union down to the remaining type(s) when one of them is rejected in other contexts  (e.g. if you `assert x is not None`, it can tell that `x` is `T` in `Optional[T`]).  But I guess there's not much value in trying to figure out why it couldn't do that here.

timj replied:
I think adding 4 asserts (two in each if branch) might do it but I'm not sure if that's much of an improvement.

TallJimbo replied:
Agreed; fine as-is.

TallJimbo reviewed (approved):
As an aside: it'd be really nice to be able to click a button on all those CodeCov annotations and say, "don't worry about coverage for this line ever again".  I don't suppose you know of anything like that?

timj commented:
The code coverage reports only turn up for new code on this pull request. If I merge this PR and no-one else ever touches that code, codecov will never report any coverage problem. I added new code with an if clause that is not triggered by any of the tests and so that counts against this PR as being a drop in coverage.  Codecov is complaining about this PR because there was a net decrease in code coverage because I added all these checks for mypy. I can ignore them and merge or I can create a few bad config files and try to create butlers from them.

TallJimbo commented:
> I can ignore them and merge or I can create a few bad config files and try to create butlers from them.

I am all for ignoring them here.  I was just hoping we could ignore them forever (e.g. keep them from coming back again if we, say, `black`-formatted everything), but it sounds like it doesn't matter much anyway.

timj commented:
If we black-formatted everything then codecov would absolutely lose its mind because of the 16% of the codebase that has no code coverage at all. We'd have to take that on the chin. I don't think it's unreasonable for codecov to warn you when you've edited a line but don't have any test code to show that the edit made a difference.

timj committed:
lsst.utils.iteration.iterable is now ensure_iterable

timj committed:
Rearrange the pixelization class logic to appease mypy

timj committed:
Add missing news fragment for DM-31976

timj committed:
Add news fragment
